File: yologram-api-v2/app/domain/ums/user_password_reset_service.py
import random
from datetime import datetime, timedelta

import bcrypt
from sqlalchemy.orm import Session

from app.core.exception import (
    UserPasswordResetExpiredException,
    UserPasswordResetInvalidException,
    UserNotFoundException,
)
from app.domain.ums.email_sender import EmailSender
from app.domain.ums.model import UserPasswordResetCode
from app.domain.ums.repository import UserPasswordResetCodeRepository, UserRepository


class UserPasswordResetService:

    def __init__(self, db: Session, email_sender: EmailSender):
        self.repository = UserPasswordResetCodeRepository(db)
        self.user_repository = UserRepository(db)
        self.email_sender = email_sender

    def send_code(self, email: str) -> None:
        if not self.user_repository.find_by_email(email):
            raise UserNotFoundException()

        self.repository.delete_by_email(email)

        code = f"{random.randint(0, 999999):06d}"
        entity = UserPasswordResetCode(
            email=email,
            code=code,
            expired_at=datetime.now() + timedelta(minutes=5),
        )
        self.repository.save(entity)
        self.email_sender.send_password_reset_code(email, code)

    def verify_code(self, email: str, code: str) -> None:
        entity = self._find_valid_code(email, code)
        entity.verified = True

    def confirm(self, email: str, code: str, new_password: str) -> None:
        self._find_valid_code(email, code)

        user = self.user_repository.find_by_email(email)
        if not user:
            raise UserNotFoundException()

        user.password = bcrypt.hashpw(
            new_password.encode("utf-8"), bcrypt.gensalt()
        ).decode("utf-8")

        self.repository.delete_by_email(email)

    def _find_valid_code(self, email: str, code: str) -> UserPasswordResetCode:
        entity = self.repository.find_latest_by_email(email)
        if not entity:
            raise UserPasswordResetInvalidException()
        if entity.expired_at < datetime.now():
            raise UserPasswordResetExpiredException()
        if entity.code != code:
            raise UserPasswordResetInvalidException()
        return entity

File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/domain/ums/service/UserPasswordResetService.kt
package link.yologram.api.v1.domain.ums.service

import link.yologram.api.v1.domain.ums.entity.UserPasswordResetCode
import link.yologram.api.v1.domain.ums.exception.UserPasswordResetExpiredException
import link.yologram.api.v1.domain.ums.exception.UserPasswordResetInvalidException
import link.yologram.api.v1.domain.ums.exception.UserNotFoundException
import link.yologram.api.v1.domain.ums.repository.UserPasswordResetCodeRepository
import link.yologram.api.v1.domain.ums.repository.UserRepository
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import kotlin.random.Random

@Service
class UserPasswordResetService(
    private val passwordResetCodeRepository: UserPasswordResetCodeRepository,
    private val userRepository: UserRepository,
    private val emailSender: EmailSender,
    private val passwordEncoder: BCryptPasswordEncoder,
) {

    @Transactional
    fun sendCode(email: String) {
        if (!userRepository.existsByEmail(email)) {
            throw UserNotFoundException()
        }

        passwordResetCodeRepository.deleteAllByEmail(email)

        val code = generateCode()
        val resetCode = UserPasswordResetCode(
            email = email,
            code = code,
            expiredAt = LocalDateTime.now().plusMinutes(5),
        )
        passwordResetCodeRepository.save(resetCode)

        emailSender.sendUserPasswordResetCode(email, code)
    }

    @Transactional
    fun verifyCode(email: String, code: String) {
        val resetCode = findValidCode(email, code)
        resetCode.verified = true
    }

    @Transactional
    fun confirm(email: String, code: String, newPassword: String) {
        findValidCode(email, code)

        val user = userRepository.findByEmail(email)
            .orElseThrow { UserNotFoundException() }
        user.password = passwordEncoder.encode(newPassword)

        passwordResetCodeRepository.deleteAllByEmail(email)
    }

    private fun findValidCode(email: String, code: String): UserPasswordResetCode {
        val resetCode = passwordResetCodeRepository.findTopByEmailOrderByCreatedAtDesc(email)
            .orElseThrow { UserPasswordResetInvalidException() }

        if (resetCode.expiredAt.isBefore(LocalDateTime.now())) {
            throw UserPasswordResetExpiredException()
        }

        if (resetCode.code != code) {
            throw UserPasswordResetInvalidException()
        }

        return resetCode
    }

    private fun generateCode(): String {
        return String.format("%06d", Random.nextInt(1_000_000))
    }
}
